test/s3select_test_json_parser.cpp

```diff
@@ -0,0 +1,586 @@
+
+	if (pattern.size()) {
+		handler.filter_exact = pattern;
+		status = handler.process_json_buffer(base64_decode(std::string(in)).data(), strlen(in));
+	}
+
+	if(status==0)
+	{
+		return handler.get_filter_result();	
+	}
+
+	return std::string("failure-sax");
+}
+
+std::string run_prefix_suffix(const char * in)
+{
+	JsonParserHandler handler;
+
+	handler.prefix_suffix_exact_criteria = true;
+
+	handler.prefix = {"nested_obj", "nested2"};
+
+	handler.suffix = {"c1"};
+
+	handler.filter_prefix_suffix.reserve(handler.prefix.size() + handler.suffix.size());
+	handler.filter_prefix_suffix.insert(handler.filter_prefix_suffix.end(), handler.prefix.begin(), handler.prefix.end());
+	handler.filter_prefix_suffix.insert(handler.filter_prefix_suffix.end(), handler.suffix.begin(), handler.suffix.end());
+
+	int status = handler.process_json_buffer(base64_decode(std::string(in)).data(), strlen(in));
+
+	if(status==0)
+	{
+		return handler.get_filter_result();	
+	}
+
+	return std::string("failure-sax");
+}
+
+std::string run_dom(const char * in)
+{
+	rapidjson::Document document;
+	document.Parse( base64_decode(std::string(in)).data() );
+
+	if (document.HasParseError()) {
+		std::cout<<"parsing error-dom"<< std::endl;
+		return std::string("parsing error");
+	}
+
+	if (!document.IsObject())
+	{
+		std::cout << " input is not an object dom" << std::endl;
+		return std::string("object error");
+	}
+
+	dom_traverse_v2 td2;
+	td2.traverse( document );
+	return std::string( (td2.ss).str() );
+}
+
+int compare_results(const char *in)
+{
+	std::cout << "===" << std::endl << base64_decode(std::string(in)) << std::endl;
+
+	std::string dom_res = run_dom(in);
+	std::string sax_res = run_sax(in);
+
+	std::cout<<"sax res is "<<sax_res<<"\n";
+
+	std::cout<<"dom res is "<<dom_res<<"\n";
+
+	auto res = dom_res.compare(sax_res);
+
+	std::cout << "dom = sax compare is :" << res << std::endl;
+	//std::cout << "key-value:\n" << sax_res << std::endl;
+
+	return res;
+}
+
+int sax_filter_results(const char *in, std::vector<std::string>& pattern)
+{
+	std::cout << "===" << std::endl << base64_decode(std::string(in)) << std::endl;
+
+	std::string sax_res = run_filter_sax(in, pattern);
+
+	std::cout << "filter result is " << sax_res << std::endl;
+
+	return 0;
+}
+
+int sax_exact_filter(const char *in, std::vector<std::string>& from_clause, const std::vector<std::string>& projection = {}, const std::vector<std::string>& where_clause = {})
+{
+	std::vector<std::string> filter_pattern1{};
+	std::vector<std::string> filter_pattern{};
+	std::string sax_res{};
+	std::string sax_res1{};
+	if (projection.size()) {
+    filter_pattern1.reserve(from_clause.size() + projection.size());
+	filter_pattern1.insert(filter_pattern1.end(), from_clause.begin(), from_clause.end());
+	filter_pattern1.insert(filter_pattern1.end(), projection.begin(), projection.end());
+	sax_res1 = run_exact_filter(in, filter_pattern1);
```

Comment:
upon a JSON query, the projections and where-clause create **many** variable names, which are actually exact-filters.
the JSON handler is missing the interface to get many exact-filters,
and it's also missing the implementation for filetring many exact-filters.

as a side note, the from-clause creates a patterns that filter out the keys that do not match the from-clause,
it's  a kind of optimization, we can deal with that later.

Reply:
I have used only 1 projection and where clause per query. Will update that.